scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamJoiner.scala

```diff
@@ -0,0 +1,283 @@
+      extends ((K, Iterable[V1], Iterable[V2]) => Iterable[R])
+
+  final case class InnerJoin[K, V1, V2]() extends JoinFunction[K, V1, V2, (V1, V2)] {
+    def apply(k: K, left: Iterable[V1], right: Iterable[V2]): Iterable[(V1, V2)] =
+      right.flatMap(v2 => left.map((_, v2)))
+  }
+  final case class LeftJoin[K, V1, V2]() extends JoinFunction[K, V1, V2, (V1, Option[V2])] {
+    def apply(k: K, left: Iterable[V1], right: Iterable[V2]): Iterable[(V1, Option[V2])] =
+      asOuter(right).flatMap(v2 => left.map((_, v2)))
+  }
+  final case class RightJoin[K, V1, V2]() extends JoinFunction[K, V1, V2, (Option[V1], V2)] {
+    def apply(k: K, left: Iterable[V1], right: Iterable[V2]): Iterable[(Option[V1], V2)] =
+      right.flatMap(v2 => asOuter(left).map((_, v2)))
+  }
+  final case class OuterJoin[K, V1, V2]() extends JoinFunction[K, V1, V2, (Option[V1], Option[V2])] {
+    def apply(k: K, left: Iterable[V1], right: Iterable[V2]): Iterable[(Option[V1], Option[V2])] =
+      if (left.isEmpty && right.isEmpty) Iterable.empty
+      else asOuter(right).flatMap(v2 => asOuter(left).map((_, v2)))
+  }
+  final case class FilteredJoin[K, V1, V2, R](jf: JoinFunction[K, V1, V2, R], fn: ((K, R)) => Boolean)
+      extends JoinFunction[K, V1, V2, R] {
+    def apply(k: K, left: Iterable[V1], right: Iterable[V2]): Iterable[R] =
+      jf.apply(k, left, right).filter(r => fn((k, r)))
+  }
+  final case class MappedJoin[K, V1, V2, R, R1](jf: JoinFunction[K, V1, V2, R], fn: R => R1)
+      extends JoinFunction[K, V1, V2, R1] {
+    def apply(k: K, left: Iterable[V1], right: Iterable[V2]): Iterable[R1] =
+      jf.apply(k, left, right).map(fn)
+  }
+  final case class FlatMappedJoin[K, V1, V2, R, R1](
+      jf: JoinFunction[K, V1, V2, R],
+      fn: R => TraversableOnce[R1]
+  ) extends JoinFunction[K, V1, V2, R1] {
+    def apply(k: K, left: Iterable[V1], right: Iterable[V2]): Iterable[R1] =
+      jf.apply(k, left, right).flatMap(fn)
+  }
+  final case class MappedGroupJoin[K, V1, V2, R, R1](
+      jf: JoinFunction[K, V1, V2, R],
+      fn: (K, Iterable[R]) => Iterable[R1]
+  ) extends JoinFunction[K, V1, V2, R1] {
+    def apply(k: K, left: Iterable[V1], right: Iterable[V2]): Iterable[R1] = {
+      val iterr = jf.apply(k, left, right)
+      if (iterr.isEmpty) Iterable.empty // mapGroup operates on non-empty groups
+      else fn(k, iterr)
+    }
+  }
+  final case class JoinFromHashJoin[K, V1, V2, R](hj: (K, V1, Iterable[V2]) => Iterable[R])
+      extends JoinFunction[K, V1, V2, R] {
+    def apply(k: K, left: Iterable[V1], right: Iterable[V2]): Iterable[R] =
+      left.flatMap(hj(k, _, right))
+  }
+
+  final case class ArbitraryJoin[K, V1, V2, R](fn: (K, Iterator[V1], Iterable[V2]) => Iterator[R])
+      extends JoinFunction[K, V1, V2, R] {
+    def apply(k: K, left: Iterable[V1], right: Iterable[V2]): Iterable[R] =
+      new Iterable[R] {
+        def iterator: Iterator[R] = fn(k, left.iterator, right)
```

Comment:
same comment about how the `Iterable` is used. If someone just calls `.iterator` repeatedly, this should be fine.

But if that's the case, returning `fn(k, left.iterator, right).toStream` would also be okay. Stream has the advantage of being lazy on filter, map, flatMap.